patches/usb-gadget-interval-hid.patch

```diff
@@ -0,0 +1,152 @@
+ };
+ 
+ static struct usb_ss_ep_comp_descriptor hidg_ss_out_comp_desc = {
+@@ -157,10 +152,7 @@ static struct usb_endpoint_descriptor hidg_hs_in_ep_desc = {
+ 	.bEndpointAddress	= USB_DIR_IN,
+ 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
+ 	/*.wMaxPacketSize	= DYNAMIC */
+-	.bInterval		= 4, /* FIXME: Add this field in the
+-				      * HID gadget configuration?
+-				      * (struct hidg_func_descriptor)
+-				      */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_endpoint_descriptor hidg_hs_out_ep_desc = {
+@@ -169,10 +161,7 @@ static struct usb_endpoint_descriptor hidg_hs_out_ep_desc = {
+ 	.bEndpointAddress	= USB_DIR_OUT,
+ 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
+ 	/*.wMaxPacketSize	= DYNAMIC */
+-	.bInterval		= 4, /* FIXME: Add this field in the
+-				      * HID gadget configuration?
+-				      * (struct hidg_func_descriptor)
+-				      */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_descriptor_header *hidg_hs_descriptors[] = {
+@@ -191,10 +180,7 @@ static struct usb_endpoint_descriptor hidg_fs_in_ep_desc = {
+ 	.bEndpointAddress	= USB_DIR_IN,
+ 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
+ 	/*.wMaxPacketSize	= DYNAMIC */
+-	.bInterval		= 10, /* FIXME: Add this field in the
+-				       * HID gadget configuration?
+-				       * (struct hidg_func_descriptor)
+-				       */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_endpoint_descriptor hidg_fs_out_ep_desc = {
+@@ -203,10 +189,7 @@ static struct usb_endpoint_descriptor hidg_fs_out_ep_desc = {
+ 	.bEndpointAddress	= USB_DIR_OUT,
+ 	.bmAttributes		= USB_ENDPOINT_XFER_INT,
+ 	/*.wMaxPacketSize	= DYNAMIC */
+-	.bInterval		= 10, /* FIXME: Add this field in the
+-				       * HID gadget configuration?
+-				       * (struct hidg_func_descriptor)
+-				       */
++	/*.bInterval		= DYNAMIC */
+ };
+ 
+ static struct usb_descriptor_header *hidg_fs_descriptors[] = {
+@@ -749,6 +732,7 @@ static int hidg_bind(struct usb_configuration *c, struct usb_function *f)
+ 	struct device		*device;
+ 	int			status;
+ 	dev_t			dev;
++	unsigned short interval_ms = cpu_to_le16(hidg->interval);
+ 
+ 	/* maybe allocate device-global string IDs, and patch descriptors */
+ 	us = usb_gstrings_attach(c->cdev, ct_func_strings,
+@@ -779,16 +763,22 @@ static int hidg_bind(struct usb_configuration *c, struct usb_function *f)
+ 	hidg_interface_desc.bInterfaceSubClass = hidg->bInterfaceSubClass;
+ 	hidg_interface_desc.bInterfaceProtocol = hidg->bInterfaceProtocol;
+ 	hidg->protocol = HID_REPORT_PROTOCOL;
++	hidg_ss_in_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_ss_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
+ 	hidg_ss_in_comp_desc.wBytesPerInterval =
+ 				cpu_to_le16(hidg->report_length);
+ 	hidg_hs_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_hs_in_ep_desc.bInterval = USB_MS_TO_HS_INTERVAL(interval_ms);
+ 	hidg_fs_in_ep_desc.wMaxPacketSize = cpu_to_le16(hidg->report_length);
++	hidg_fs_in_ep_desc.bInterval = interval_ms;
```

Comment:
is this intentional that the same variable is assigned both?
- `USB_MS_TO_HS_INTERVAL(interval_ms)`
- `interval_ms`

Reply:
it's normal:

- Full Speed interval can be taken as it is (in ms)

```
	usb_ep_desc_binterval_fs_1_ms	= 1,
	usb_ep_desc_binterval_fs_2_ms	= 2,
	usb_ep_desc_binterval_fs_4_ms	= 4,
	usb_ep_desc_binterval_fs_8_ms	= 8,
	usb_ep_desc_binterval_fs_16_ms	= 16,
	usb_ep_desc_binterval_fs_32_ms	= 32,
```

- SuperSpeed and HighSpeed need to be converted.

```
	usb_ep_desc_binterval_hs_125_us	= 1,
	usb_ep_desc_binterval_hs_250_us	= 2,
	usb_ep_desc_binterval_hs_500_us	= 3,
	usb_ep_desc_binterval_hs_1_ms	= 4,
	usb_ep_desc_binterval_hs_2_ms	= 5,
	usb_ep_desc_binterval_hs_4_ms	= 6
```